Wrappers/Python/ccpi/viewer/CILViewer.py

```diff
@@ -827,30 +1053,27 @@ def updatePipeline(self, resetcamera = False):
 
     def updateVolumePipeline(self):
         if self.volume_render_initialised and self.volume.GetVisibility():
-            cmin , cmax = self.volume_colormap_limits
-            colors = colormaps.CILColorMaps.get_color_transfer_function(self.volume_colormap_name, (cmin,cmax))
-
-            x = numpy.linspace(self.ia.GetMinimum(), self.ia.GetMaximum(), num=255)
-            scaling = 0.1
-            opacity = colormaps.CILColorMaps.get_opacity_transfer_function(x, 
-            colormaps.relu, cmin, cmax, scaling)
+            # define colors and opacity with default values
+            colors, opacity = self.getColorOpacityForVolumeRender()
+            
             self.volume_property.SetColor(colors)
-            self.volume_property.SetScalarOpacity(opacity)
-        
-
-    def setVolumeColorLevelWindow(self, cmin, cmax):
-        self.volume_colormap_limits = (cmin, cmax)
-        self.updatePipeline()
 
-    def setVolumeColorName(self, name):
-        self.volume_colormap_name = name
-        self.updatePipeline()
+            # Update whether we use our calculated opacity as the scalar or gradient opacity
+            if self.getVolumeRenderOpacityMethod() == 'gradient':
+                # Also return the scalar opacity to its default value:
+                self.volume_property.SetScalarOpacity(self._getDefaultScalarOpacityFunction()) 
+                self.volume_property.DisableGradientOpacityOff()
+                self.volume_property.SetGradientOpacity(opacity)
+                 
+            elif self.getVolumeRenderOpacityMethod() == 'scalar': 
+                self.volume_property.DisableGradientOpacityOn()
+                self.volume_property.SetScalarOpacity(opacity)
```

Comment:
Do I understand correctly that you don't have to set both scalar and gradient opacity as you found before?

Reply:
If we want to use the scalar opacity, we don't have to use a gradient opacity - hence we disable it.
If we want to use a gradient opacity, there still exists a scalar opacity. Currently what happens is if we switch to gradient it uses the default scalar opacity from vtk. Maybe instead we should use *our* default scalar opacity, or whatever scalar opacity the user has set

Reply:
added comments to code to explain why we do this